Add tracks event for WooCommerce Payments install via wc-addons page

chickenn00dle opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:
This adds a tracks event for WooCommerce Payments installs via the
wc-addons page.

See p1612278892009000-slack-C0144BMA170 for more details.

### How to test the changes in this Pull Request:

1. Make sure your store location is set to within the US and WooCommerce Payments is not installed.
2. Go to the wc-addons page (**WooCommerce > Home > Extensions**) and look for the following banner: https://cloudup.com/c2tYAq7O4Wp
3. Install via the button, and after 5-10 minutes check tracks live view for a `wcadmin_woocommerce_payments_install` with the context prop of `extensions`.

Note this event is also implemented in https://github.com/woocommerce/woocommerce-admin/pull/6285 with the context of `tasklist`.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Dev - Track the number of installations of WooCommerce Payments via that extensions banner for stores that have opted in to tracking.

chickenn00dle committed:
Add tracks event for WCPay install via addons page

This adds a tracks event for WooCommerce Payments installs via the
wc-addons page.

vbelolapotkov commented on includes/tracks/events/class-wc-extensions-tracking.php:
```diff
@@ -21,6 +21,7 @@ public function init() {
 		add_action( 'woocommerce_helper_connected', array( $this, 'track_helper_connection_complete' ) );
 		add_action( 'woocommerce_helper_disconnected', array( $this, 'track_helper_disconnected' ) );
 		add_action( 'woocommerce_helper_subscriptions_refresh', array( $this, 'track_helper_subscriptions_refresh' ) );
+		add_action( 'wc_addons_woocommerce_payments_installed', array( $this, 'track_woocommerce_payments_install' ) );
```
Not sure about hook name, it looks like all hooks above have `woocommerce_` prefix rather than `wc_`. Sine there were a request to have similar tracking for other extensions like WC Ship, would it make sense to have a shared action like `woocommerce_addon_installed` which would accept addon name as an argument.

Then it'd be called like `do_action( 'woocommerce_addon_installed', 'woocommerce_payments' );`

chickenn00dle replied:
Good call! I've made this change in https://github.com/woocommerce/woocommerce/pull/29052/commits/b999ae51839c8760023c5d8296e347155e62160b

Just noting that in order to follow the tracks naming conventions as closely as possible, I didn't programmatically create the event name using `$addon_id`:

> Don’t programmatically create event names. They should be hardcoded so they are easily searchable in the code base.

vbelolapotkov replied:
> I didn't programmatically create the event name using $addon_id

💯 that's right approach.

vbelolapotkov commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -704,7 +704,9 @@ public static function install_woocommerce_payments_addon() {
-		WC_Install::background_installer( $services_plugin_id, $wcpay_plugin );
+		WC_Install::background_installer( $wcpay_plugin_id, $wcpay_plugin );
+
+		do_action( 'wc_addons_woocommerce_payments_installed' );
```
Can we find out here that extension was installed from `Payments` tab of marketplace or other tabs: e.g. `Featured`, `All`, etc?

chickenn00dle replied:
Hey @vbelolapotkov 

This would be possible. The tab is available via the URL `section` parameter (if this doesn't exist, then the tab is `_featured`). Do you think it's worth distinguishing between the two in our event? I wasn't sure if we needed to be this granular. 

Its worth noting that the banner only appears to display in the featured and payments tabs.

vbelolapotkov replied:
I don't have answer here, so asked in slack :)

chickenn00dle replied:
We got an answer in Slack and I added the section to props in https://github.com/woocommerce/woocommerce/pull/29052/commits/7a4b54ad38cad180e6353b936b080774b25a50a4

chickenn00dle committed:
Change hook name and add id param to track method

This changes the addon install hook to match the other hooks in the
extensions tracking class. This also adds an addon_id param to the
callback associated with this hook so we can extend this to other
plugins in the future.

vbelolapotkov commented on includes/tracks/events/class-wc-extensions-tracking.php:
```diff
@@ -76,4 +77,21 @@ public function track_helper_disconnected() {
+	 * @param string $addon_id Addon slug.
+	 */
+	public function track_addon_install( $addon_id = '' ) {
+		if ( empty( $addon_id ) ) {
+			return;
+		}
+
+		$properties = array( 'context' => 'extensions' );
+
+		if ( 'woocommerce-payments' === $addon_id ) {
+			WC_Tracks::record_event( 'woocommerce_payments_install', $properties );
+		}
```
Not critical, but switch/case would look better here IMO:
```suggestion
		switch ( $addon_id ) {
		    case 'woocommerce-payments':
		        WC_Tracks::record_event( 'woocommerce_payments_install', array( 'context' => 'extensions' ) );
		        break;
		}
```

chickenn00dle replied:
I thought about this but didn't want to go that route until tracking for other addons were added here.

vbelolapotkov replied:
Ok, no problem, any solution would work for start :)

chickenn00dle committed:
Add section to tracks event props

This adds the Extenstions tab to the event props so we can determine
which tab merchants are installing wcpay from.

vbelolapotkov reviewed (approved):
LGTM!

chickenn00dle commented:
Awesome! I'm not authorized to merge this one, unfortunately? Is this something you can do @vbelolapotkov?

peterfabian commented:
We need to review as well and can merge afterward. Let me have a look.

peterfabian reviewed (approved):
Seems to work fine, thanks.
